### PR TITLE
Azure Batch 2022-01-01.15.0 Data Plane Update

### PR DESCRIPTION
**Pull request by paterasMSFT**
- Added property uploadHeaders to OutputFileBlobContainerDestination.
  - Allows users to set custom HTTP headers on resource file uploads.
  - Array of type HttpHeader (also being added).
- Added boolean property allowTaskPreemption to JobSpecification, CloudJob, JobAddParameter, JobPatchParameter, JobUpdateParameter
  - Mark Tasks as preemptible for higher priority Tasks (requires Comms-Enabled or Single Tenant Pool).
- Replaced comment (title, description, etc.) references of "low-priority" with "Spot/Low-Priority", to reflect new service behavior.
  - No API change required.
  - Low-Priority Compute Nodes (VMs) will continue to be used for User Subscription pools (and only User Subscription pools), as before.
  - Spot Compute Nodes (VMs) will now be used for Batch Managed (and only Batch Managed pools) pools.
  - Relevant docs:
    - https://docs.microsoft.com/en-us/azure/batch/nodes-and-pools
    - https://docs.microsoft.com/en-us/azure/batch/batch-spot-vms

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [x] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [x] SDK of Java
      - [x] SDK of Js
      - [x] SDK of Go
      - [x] PowerShell
      - [x] CLI
      - [x] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[x] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 1]
```diff
@@ -11773,6 +11773,22 @@
...
+        },
+        "value": {
+          "type": "string",
+          "title": "The value of the header to be used while uploading output files"
+        }
+      },
+      "required": [
+        "name"
+      ],
+      "title": "A HTTP header name-value pair"
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:11790]: Minor grammar fix: "**An** HTTP header name-value pair"

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 2]
```diff
@@ -12670,12 +12686,17 @@
           "title": "The priority of Jobs created under this schedule.",
           "description": "Priority values can range from -1000 to 1000, with -1000 being the lowest priority and 1000 being the highest priority. The default value is 0. This priority is used as the default for all Jobs under the Job Schedule. You can update a Job's priority after it has been created using by using the update Job API."
         },
+        "allowTaskPreemption": {
+          "type": "boolean",
+          "title": "Whether Tasks in this job can be preempted by other high priority jobs",
+          "description": "If the value is set to True other high priority jobs submitted to the system will take precedence and will be able requeue tasks from this job. You can update a job's allowTaskPreemption after it has been created using the update job API."
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:12692]: Missing comma: "If the value is set to True,"

Comment (dpwatrous) [reply specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:12692]: This is a bit out of scope of this PR as it affects other properties too, but it seems weird to call out every updatable property in the description individually.

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 3]
```diff
@@ -13135,12 +13156,17 @@
           "title": "The priority of the Job.",
           "description": "Priority values can range from -1000 to 1000, with -1000 being the lowest priority and 1000 being the highest priority. The default value is 0."
         },
+        "allowTaskPreemption": {
+          "type": "boolean",
+          "title": "Whether Tasks in this job can be preempted by other high priority jobs",
+          "description": "If the value is set to True other high priority jobs submitted to the system will take precedence and will be able requeue tasks from this job. You can update a job's allowTaskPreemption after it has been created using the update job API."
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:13162]: Missing comma: "If the value is set to True,"

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 4]
```diff
@@ -13230,7 +13256,12 @@
...
-          "description": "The value of maxParallelTasks must be -1 or greater than 0 if specified. If not specified, the default value is -1, which means there's no limit to the number of tasks that can be run at once. You can update a job's maxParallelTasks after it has been created using the update job API. This property is currently in private preview."
+          "description": "The value of maxParallelTasks must be -1 or greater than 0 if specified. If not specified, the default value is -1, which means there's no limit to the number of tasks that can be run at once. You can update a job's maxParallelTasks after it has been created using the update job API."
+        },
+        "allowTaskPreemption": {
+          "type": "boolean",
+          "title": "Whether Tasks in this job can be preempted by other high priority jobs",
+          "description": "If the value is set to True other high priority jobs submitted to the system will take precedence and will be able requeue tasks from this job. You can update a job's allowTaskPreemption after it has been created using the update job API."
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:13264]: Missing comma: "If the value is set to True,"

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 5]
```diff
@@ -14496,7 +14527,7 @@
           "type": "integer",
           "format": "int32",
           "title": "The number of low-priority Compute Nodes currently in the Pool.",
-          "description": "low-priority Compute Nodes which have been preempted are included in this count."
+          "description": "Low-priority Compute Nodes which have been preempted are included in this count."
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:14530]: Let's do a find/replace on Low-priority and change to "Spot/low-priority" (with Spot capitalized and low-priority lowercase to match other docs)

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 6]
```diff
@@ -16214,7 +16245,12 @@
...
-          "description": "The value of maxParallelTasks must be -1 or greater than 0 if specified. If not specified, the default value is -1, which means there's no limit to the number of tasks that can be run at once. You can update a job's maxParallelTasks after it has been created using the update job API. This property is currently in private preview."
+          "description": "The value of maxParallelTasks must be -1 or greater than 0 if specified. If not specified, the default value is -1, which means there's no limit to the number of tasks that can be run at once. You can update a job's maxParallelTasks after it has been created using the update job API."
+        },
+        "allowTaskPreemption": {
+          "type": "boolean",
+          "title": "Whether Tasks in this job can be preempted by other high priority jobs",
+          "description": "If the value is set to True other high priority jobs submitted to the system will take precedence and will be able requeue tasks from this job. You can update a job's allowTaskPreemption after it has been created using the update job API."
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:16253]: Missing comma: "If the value is set to True,"

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 7]
```diff
@@ -16251,6 +16287,18 @@
...
+          "type": "integer",
+          "format": "int32",
+          "default": -1,
+          "title": "The maximum number of tasks that can be executed in parallel for the job.",
+          "description": "The value of maxParallelTasks must be -1 or greater than 0 if specified. If not specified, the default value is -1, which means there's no limit to the number of tasks that can be run at once. You can update a job's maxParallelTasks after it has been created using the update job API."
+        },
+        "allowTaskPreemption": {
+          "type": "boolean",
+          "title": "Whether Tasks in this job can be preempted by other high priority jobs",
+          "description": "If the value is set to True other high priority jobs submitted to the system will take precedence and will be able requeue tasks from this job. You can update a job's allowTaskPreemption after it has been created using the update job API."
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:16300]: Missing comma: "If the value is set to True,"

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 8]
```diff
@@ -16644,6 +16692,14 @@
...
           "description": "The identity must have write access to the Azure Blob Storage container",
           "$ref": "#/definitions/ComputeNodeIdentityReference"
+        },
+        "uploadHeaders": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/HttpHeader"
+          },
+          "title": "A list of name-value pairs for headers to be used in uploading output files",
+          "description": "These headers will be specified when uploading files to Azure Storage. Official document on allowed headers when uploading blobs: https://docs.microsoft.com/en-us/rest/api/storageservices/put-blob#request-headers-all-blob-types"
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:16702]: Wording suggestion: "Official document on allowed headers" -> "For more information, see [Request Headers (All Blob Types)](https://docs.microsoft.com/rest/api/storageservices/put-blob#request-headers-all-blob-types)."

Comment (dpwatrous) [reply specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:16702]: (This is just following a pattern I see for other swagger files linking to docs)

File: specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json [hunk 9]
```diff
@@ -17176,7 +17232,7 @@
         },
         "lowPriority": {
           "$ref": "#/definitions/NodeCounts",
-          "title": "The number of low-priority Compute Nodes in each state."
+          "title": "The number of low priority Compute Nodes in each state."
```
Comment (dpwatrous) [inline specification/batch/data-plane/Microsoft.Batch/stable/2022-01-01.15.0/BatchService.json:17235]: Not sure why we made this change as our docs use "low-priority" with a hyphen. That said, see my other comment on using "Spot/low-priority" which is what we landed on for spot support.

**Review by dpwatrous — CHANGES_REQUESTED** [review]

**Review by dpwatrous — APPROVED** [review]
Changes look good to me - thanks!

**Comment by paterasMSFT** [comment]
@anuchandy mind signing off so we can merge? Shouldn't be any ARM changes here.

**Comment by paterasMSFT** [comment]
@anuchandy would you mind reviewing when you get a chance? This PR is ready to merge (the azure-sdsk-for-js failure is expected and not an issue).

**Review by lmazuel — APPROVED** [review]
Was approved by Mark and Mike offline

**Pull request merged**